diff --git a/BaseTools/Source/Python/AutoGen/ModuleAutoGen.py b/BaseTools/Source/Python/AutoGen/ModuleAutoGen.py
index f610365f6d..b48d7027de 100755
--- a/BaseTools/Source/Python/AutoGen/ModuleAutoGen.py
+++ b/BaseTools/Source/Python/AutoGen/ModuleAutoGen.py
@@ -1802,17 +1802,21 @@ class ModuleAutoGen(AutoGen):
 
         MewIR = ModuleBuildCacheIR(self.MetaFile.Path, self.Arch)
         MewIR.MakefilePath = MakefilePath
-        MewIR.DependencyHeaderFileSet = Makefile.DependencyHeaderFileSet
+        if len(self.CustomMakefile) == 0:
+            MewIR.DependencyHeaderFileSet = Makefile.DependencyHeaderFileSet
         MewIR.CreateMakeFileDone = True
-        with GlobalData.cache_lock:
-            try:
-                IR = gDict[(self.MetaFile.Path, self.Arch)]
-                IR.MakefilePath = MakefilePath
-                IR.DependencyHeaderFileSet = Makefile.DependencyHeaderFileSet
-                IR.CreateMakeFileDone = True
-                gDict[(self.MetaFile.Path, self.Arch)] = IR
-            except:
-                gDict[(self.MetaFile.Path, self.Arch)] = MewIR
+        if GlobalData.cache_lock:
+            with GlobalData.cache_lock:
+                try:
+                    IR = gDict[(self.MetaFile.Path, self.Arch)]
+                    IR.MakefilePath = MakefilePath
+                    IR.DependencyHeaderFileSet = Makefile.DependencyHeaderFileSet
+                    IR.CreateMakeFileDone = True
+                    gDict[(self.MetaFile.Path, self.Arch)] = IR
+                except:
+                    gDict[(self.MetaFile.Path, self.Arch)] = MewIR
+        else:
+            gDict[(self.MetaFile.Path, self.Arch)] = MewIR
 
     def CopyBinaryFiles(self):
         for File in self.Module.Binaries:
@@ -1889,13 +1893,16 @@ class ModuleAutoGen(AutoGen):
         self.IsCodeFileCreated = True
         MewIR = ModuleBuildCacheIR(self.MetaFile.Path, self.Arch)
         MewIR.CreateCodeFileDone = True
-        with GlobalData.cache_lock:
-            try:
-                IR = gDict[(self.MetaFile.Path, self.Arch)]
-                IR.CreateCodeFileDone = True
-                gDict[(self.MetaFile.Path, self.Arch)] = IR
-            except:
-                gDict[(self.MetaFile.Path, self.Arch)] = MewIR
+        if GlobalData.cache_lock:
+            with GlobalData.cache_lock:
+                try:
+                    IR = gDict[(self.MetaFile.Path, self.Arch)]
+                    IR.CreateCodeFileDone = True
+                    gDict[(self.MetaFile.Path, self.Arch)] = IR
+                except:
+                    gDict[(self.MetaFile.Path, self.Arch)] = MewIR
+        else:
+            gDict[(self.MetaFile.Path, self.Arch)] = MewIR
 
         return AutoGenList
 
